datafusion/core/src/catalog/dynamic_file_schema.rs

```diff
@@ -0,0 +1,118 @@
+use std::any::Any;
+use std::sync::{Arc, Weak};
+
+use async_trait::async_trait;
+use dirs::home_dir;
+use parking_lot::{Mutex, RwLock};
+
+use datafusion_common::plan_datafusion_err;
+
+use crate::catalog::schema::SchemaProvider;
+use crate::datasource::listing::{ListingTable, ListingTableConfig, ListingTableUrl};
+use crate::datasource::TableProvider;
+use crate::error::Result;
+use crate::execution::context::SessionState;
+
+/// Wraps another schema provider
+pub struct DynamicFileSchemaProvider {
+    inner: Arc<dyn SchemaProvider>,
+    state_store: StateStore,
+}
+
+impl DynamicFileSchemaProvider {
+    pub fn new(inner: Arc<dyn SchemaProvider>) -> Self {
+        Self {
+            inner,
+            state_store: StateStore::new(),
+        }
+    }
+
+    pub fn with_state(&self, state: Weak<RwLock<SessionState>>) {
+        self.state_store.with_state(state);
+    }
+}
+
+#[async_trait]
+impl SchemaProvider for DynamicFileSchemaProvider {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn table_names(&self) -> Vec<String> {
+        self.inner.table_names()
+    }
+
+    fn register_table(
+        &self,
+        name: String,
+        table: Arc<dyn TableProvider>,
+    ) -> Result<Option<Arc<dyn TableProvider>>> {
+        self.inner.register_table(name, table)
+    }
+
+    async fn table(&self, name: &str) -> Result<Option<Arc<dyn TableProvider>>> {
+        let inner_table = self.inner.table(name).await?;
+        if inner_table.is_some() {
+            return Ok(inner_table);
+        }
+        let optimized_url = substitute_tilde(name.to_owned());
+        let table_url = ListingTableUrl::parse(optimized_url.as_str())?;
+        let state = &self
+            .state_store
+            .get_state()
```

Comment:
`SchemaProvider::table` won't have `SessionState` by default.  We should get it from `StateStore`.